tests/scripts/thread-cert/Cert_5_1_01_RouterAttach.py

```diff
@@ -168,6 +179,317 @@ def test(self):
+                              } <= set(p.mle.tlv.type) or\
+                             {\
+                              LINK_LAYER_FRAME_COUNTER_TLV,\
+                              MODE_TLV,\
+                              RESPONSE_TLV,\
+                              TIMEOUT_TLV,\
+                              TLV_REQUEST_TLV,\
+                              ADDRESS16_TLV,\
+                              NETWORK_DATA_TLV,\
+                              VERSION_TLV\
+                               } <= set(p.mle.tlv.type)).\
+                   must_next()
+        _pkt.must_not_verify(lambda p: (ADDRESS_REGISTRATION_TLV) in p.mle.tlv.type)
+
+        # Step 5: Leader responds with a Child ID Response.
+        #         The following TLVs MUST be present in the Child ID Response:
+        #             - Address16 TLV
+        #             - Leader Data TLV
+        #             - Network Data TLV
+        #             - Source Address TLV
+        #             - Route64 TLV (if requested)
+
+        pkts.filter_wpan_src64(LEADER).\
+            filter_wpan_dst64(ROUTER).\
+            filter_mle_cmd(MLE_CHILD_ID_RESPONSE).\
+            filter(lambda p: {\
+                              ADDRESS16_TLV,\
+                              LEADER_DATA_TLV,\
+                              NETWORK_DATA_TLV,\
+                              SOURCE_ADDRESS_TLV,\
+                              ROUTE64_TLV\
+                              } <= set(p.mle.tlv.type) or\
+                             {\
+                              ADDRESS16_TLV,\
+                              LEADER_DATA_TLV,\
+                              NETWORK_DATA_TLV,\
+                              SOURCE_ADDRESS_TLV\
+                               } <= set(p.mle.tlv.type)).\
+                   must_next()
+
+        # Step 6: Router sends an Address Solicit Request.
+        #         Ensure the Address Solicit Request is properly formatted:
+        #         CoAP Request URI
+        #             coap://<leader address>:MM/a/as
+        #         CoAP Payload
+        #             - MAC Extended Address TLV
+        #             - Status TLV
+
+        _pkt = pkts.filter_wpan_src64(ROUTER).\
+            filter_wpan_dst16(LEADER_RLOC16).\
+            filter_coap_request(ADDR_SOL_URI).\
+            filter(lambda p: {\
+                              NL_MAC_EXTENDED_ADDRESS_TLV,
+                              NL_STATUS_TLV\
+                              } == set(p.coap.tlv.type)\
+                   ).\
+           must_next()
+
+        # Step 7: Leader sends an Address Solicit Response.
+        #         Ensure the Address Solicit Response is properly formatted:
+        #         CoAP Response Code
+        #             2.04 Changed
+        #         CoAP Payload
+        #             - Status TLV (value = Success)
+        #             - RLOC16 TLV
+        #             - Router Mask TLV
+
+        pkts.filter_wpan_src64(LEADER).\
```

Comment:
Generally TMF in Thread Core should use RLOC address as source, would it be better to use a filter like `filter_wpan_src16()` here?

Reply:
I recalled one issue regarding using rloc16 as the source filter: the rloc16 of device can change when it upgraded to Router or downgraded. Unlike RLOC16, the src64 is usually fixed. I think this is the reason why Test Harness also chose to use src64 for filtering. 
Currently, `pktverify` collects the RLOC16s of devices using `collect_rloc16s()`. So, the RLOC values could vary according to when this method is called during test procedure. 

Using the correct RLOC16 of each device at each step needs extra work, especially for the cases when the device could reset and reattach during the test. 

So, maybe using  `filter_wpan_src64` is easier and good enough for Test Plan?

Thoughts?

Reply:
Agree